src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs

```diff
@@ -670,10 +672,51 @@ public void LogTelemetry(string eventName, IDictionary<string, string> propertie
         #endregion
 
         #region IBuildEngine7 Members
+
         /// <summary>
         /// Enables or disables emitting a default error when a task fails without logging errors
         /// </summary>
         public bool AllowFailureWithoutError { get; set; } = false;
+
+        #endregion
+
+        #region IBuildEngine8 Members
+
+        int runningTotal = 0;
+
+        public int RequestCores(int requestedCores)
+        {
+            var rms = _host.GetComponent(BuildComponentType.TaskResourceManager) as ResourceManagerService;
+
+            int coresAcquiredBeforeMoreCoresGetAcquired = runningTotal;
+
+            var coresAcquired = rms.RequestCores(requestedCores);
+
+            runningTotal += coresAcquired;
+
+            return coresAcquired;
+        }
+
+        public void ReleaseCores(int coresToRelease)
+        {
+            var rms = _host.GetComponent(BuildComponentType.TaskResourceManager) as ResourceManagerService;
+
+            coresToRelease = Math.Min(runningTotal, coresToRelease);
```

Comment:
Wouldn't this mean you could request to release more cores than you own? I imagine that could have rather disastrous effects.

Reply:
It would be good to log something in diagnostic logging if a task releases more cores than it has requested.

Reply:
> I imagine that could have rather disastrous effects.

What disastrous effects? You can request whatever you want and this reduces it to what you currently have.

Reply:
I was imagining something like:
100 processes each request 1 core.
1 process releases 100 cores.
Repeat.

That may sound malicious, but if the last task requested 100 cores but got one, it might not realize it in the "release" part. If the execution as a whole is long-lived, this could promote inefficiencies.

Reply:
This is the code that prevents that: you can only release up to the current number of cores you had, as expressed by this `Math.Min` operation.